File: 0257-binary-tree-paths/solution.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    List<String> ans = new ArrayList<>();
    String tree = "";
    List<String> back = new ArrayList<>();
    
    public List<String> binaryTreePaths(TreeNode root) {
        if(root == null) return ans;

        if(root.left == null && root.right == null){
            tree += root.val;
            ans.add(tree);
        }    
        else tree += root.val + "->";
        back.add(tree);

        binaryTreePaths(root.left);
        binaryTreePaths(root.right);

        if(back.size()>0){
            back.remove(back.size()-1);
            if(!back.isEmpty())tree = back.get(back.size() - 1);
        }return ans;
    }
}

File: 0257-binary-tree-paths/solution.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    vector<string> binaryTreePaths(TreeNode* root) {
        vector<string> leafs;
        Paths(root, "", leafs);
        return leafs;
    }

    void Paths(TreeNode* root, string s, vector<string>& leafs){
        if(!root) return;
        
        s += to_string(root->val);
        if(!root->left && !root->right){
            leafs.push_back(s);
            return;
        }

        s += "->";
        Paths(root->left, s, leafs);
        Paths(root->right,s, leafs);
    }
};
